types/plugins/index.d.ts

```diff
@@ -0,0 +1,499 @@
+  title: string[];
+
+  // Footer
+  // lines of text that form the footer
+  footer: string[];
+
+  // colors to render for each item in body[]. This is the color of the squares in the tooltip
+  labelColors: Color[];
+  labelTextColors: Color[];
+
+  // 0 opacity is a hidden tooltip
+  opacity: number;
+
+  // tooltip options
+  options: ITooltipOptions;
+}
+
+export const Tooltip: IPlugin & {
+  readonly positioners: {
+    [key: string]: (items: readonly Element[], eventPosition: { x: number; y: number }) => { x: number; y: number };
+  };
+};
+
+export interface ITooltipCallbacks {
+  beforeTitle(this: TooltipModel, tooltipItems: ITooltipItem[]): string | string[];
+  title(this: TooltipModel, tooltipItems: ITooltipItem[]): string | string[];
+  afterTitle(this: TooltipModel, tooltipItems: ITooltipItem[]): string | string[];
+
+  beforeBody(this: TooltipModel, tooltipItems: ITooltipItem[]): string | string[];
+  afterBody(this: TooltipModel, tooltipItems: ITooltipItem[]): string | string[];
+
+  beforeLabel(this: TooltipModel, tooltipItem: ITooltipItem): string | string[];
+  label(this: TooltipModel, tooltipItem: ITooltipItem): string | string[];
+  afterLabel(this: TooltipModel, tooltipItem: ITooltipItem): string | string[];
+
+  labelColor(this: TooltipModel, tooltipItem: ITooltipItem): { borderColor: Color; backgroundColor: Color };
+  labelTextColor(this: TooltipModel, tooltipItem: ITooltipItem): Color;
+
+  beforeFooter(this: TooltipModel, tooltipItems: ITooltipItem[]): string | string[];
+  footer(this: TooltipModel, tooltipItems: ITooltipItem[]): string | string[];
+  afterFooter(this: TooltipModel, tooltipItems: ITooltipItem[]): string | string[];
+}
+
+export interface ITooltipPlugin<O = {}> {
+  /**
+   * @desc Called before drawing the `tooltip`. If any plugin returns `false`,
+   * the tooltip drawing is cancelled until another `render` is triggered.
+   * @param {Chart} chart - The chart instance.
+   * @param {object} args - The call arguments.
+   * @param {Tooltip} args.tooltip - The tooltip.
+   * @param {object} options - The plugin options.
+   * @returns {boolean} `false` to cancel the chart tooltip drawing.
+   */
+  beforeTooltipDraw?(chart: Chart, args: { tooltip: TooltipModel }, options: O): boolean | void;
+  /**
+   * @desc Called after drawing the `tooltip`. Note that this hook will not
+   * be called if the tooltip drawing has been previously cancelled.
+   * @param {Chart} chart - The chart instance.
+   * @param {object} args - The call arguments.
+   * @param {Tooltip} args.tooltip - The tooltip.
+   * @param {object} options - The plugin options.
+   */
+  afterTooltipDraw?(chart: Chart, args: { tooltip: TooltipModel }, options: O): void;
+}
+
+export interface ITooltipOptions {
+  /**
+   * Are on-canvas tooltips enabled?
+   * @default true
+   */
+  enabled: boolean;
+  /**
+   * 	See custom tooltip section.
+   */
+  custom(tooltip: TooltipModel, args: { chart: Chart; tooltip: TooltipModel }): void;
```

Comment:
This is called only with a single argument `args: { chart: Chart; tooltip: TooltipModel }` however the function is called such that `this` is the tooltip model. I think the correct definition would be `custom(this: TooltipModel, args: { chart: Chart; tooltip: TooltipModel }): void`